### ISSUE TITLE
optimize.reexportedNamespaces

### ISSUE THREAD

**Issue by Andarist** [rollup/rollup#3145]
### Feature Use Case

According to @lukastaegert statements on Twitter Rollup is able to drop unused exports from reexported namespaces (and webpack is supposedly going to know how to do that too).

Problem is that when bundling libraries with Rollup people most often just flat bundle and thus they convert their reexported namespaces to regular objects.

### Feature Proposal

I would like to ask for a new optimization flag that would generate a new chunk for each reexported namespace, so that bundlers don't lose that information when libraries are flat bundled (they won't be so flat bundled anymore, but still internal structure will be well hidden).

**Comment by lukastaegert** [comment]
Could you provide an actual code example with current vs. expected output of what you want to achieve? This would help make this more clear.

**Comment by Andarist** [comment]
Added to the original post - from what I understand the proposed output would help Rollup to trees-hake this:
```
import { ns } from './lib'
console.log(ns.bar)
```
so that only `bar` would be present in the final bundle (no `foo` nor `answer`).

**Comment by Andarist** [comment]
@lukastaegert does this make sense? I can close this if it doesn't 😉

**Comment by lukastaegert** [comment]
I think it does. What be a non-trivial improvement but worthwhile IMO. Only helpful for ESM output, though.

**Comment by Andarist** [comment]
> Only helpful for ESM output, though.

Yeah, that's correct.

> What be a non-trivial improvement but worthwhile IMO

Could you give me more details about this? I doubt I will have time to dive into it, but who knows... Where the complexity lies here? I would imagine that you build a graph of the modules, do tree-shaking on that and later generate the output. Wouldn't this feature be somewhat contained to the last phase (generation)?

**Comment by lukastaegert** [comment]
Yes, I hope at some point to move the chunking as a first step to the generate phase but unfortunately at the moment, it is the last step of the build phase. However, the complexity lies in how to create a namespace via an import. Ideally, the rendering itself would not need to look at the flag but always use an import to generate a namespace if the namespace corresponds to the exports of another chunk. I am on the road at the moment so cannot dive in here more.
